test/ttmlir/Silicon/TTNN/dealloc.mlir

```diff
@@ -0,0 +1,49 @@
+// RUN: ttmlir-opt --ttir-to-ttnn-backend-pipeline="system-desc-path=%system_desc_path%" %s > %t.mlir
+// RUN: FileCheck %s --input-file=%t.mlir
+// RUN: ttmlir-translate --ttnn-to-flatbuffer %t.mlir > %t.ttnn
+#any_device = #tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>
+#loc = loc("Dealloc":4294967295:0)
+module @"dealloc_test" attributes {} {
+  func.func @main(%arg0: tensor<1x784xf32> loc("Dealloc":4294967295:0), %arg1: tensor<1x10xf32> loc("Dealloc":4294967295:0), %arg2: tensor<256x10xf32> loc("Dealloc":4294967295:0), %arg3: tensor<1x256xf32> loc("Dealloc":4294967295:0), %arg4: tensor<784x256xf32> loc("Dealloc":4294967295:0)) -> tensor<1x10xf32> {
+    %0 = tensor.empty() : tensor<1x256xf32> loc(#loc8)
+    %1 = "ttir.matmul"(%arg0, %arg4, %0) <{operand_constraints = [#any_device, #any_device, #any_device]}> : (tensor<1x784xf32>, tensor<784x256xf32>, tensor<1x256xf32>) -> tensor<1x256xf32> loc(#loc8)
+    // CHECK: %{{.+}} = "ttnn.matmul"([[I1:%.+]], [[I2:%.+]], [[O1:%.+]]) {{.+}} -> tensor<1x256xf32, {{.+}}>
+    // CHECK: "ttnn.dealloc"([[I2]]) : (tensor<784x256xf32, {{.+}}) -> ()
+    // CHECK: "ttnn.dealloc"([[I1]]) : (tensor<1x784xf32, {{.+}}>) -> ()
+    %2 = tensor.empty() : tensor<1x256xf32> loc(#loc9)
+    %3 = "ttir.add"(%1, %arg3, %2) <{operandSegmentSizes = array<i32: 2, 1>, operand_constraints = [#any_device, #any_device, #any_device]}> : (tensor<1x256xf32>, tensor<1x256xf32>, tensor<1x256xf32>) -> tensor<1x256xf32> loc(#loc9)
+    // CHECK: %{{.+}} = "ttnn.add"([[I1:%.+]], [[I2:%.+]], [[O2:%.+]]) {{.+}} -> tensor<1x256xf32, {{.+}}>
+    // CHECK: "ttnn.dealloc"([[I2]]) : (tensor<1x256xf32, {{.+}}>) -> ()
+    // CHECK: "ttnn.dealloc"([[O1]]) : (tensor<1x256xf32, {{.+}}>) -> ()
+    %4 = tensor.empty() : tensor<1x256xf32> loc(#loc10)
+    %5 = "ttir.relu"(%3, %4) <{operandSegmentSizes = array<i32: 1, 1>, operand_constraints = [#any_device, #any_device]}> : (tensor<1x256xf32>, tensor<1x256xf32>) -> tensor<1x256xf32> loc(#loc10)
+    // CHECK: %{{.+}} = "ttnn.relu"([[I1:%.+]], [[O3:%.+]]) {{.+}} -> tensor<1x256xf32, {{.+}}>
+    // CHECK: "ttnn.dealloc"([[O2]]) : (tensor<1x256xf32, {{.+}}>) -> ()
+    %6 = tensor.empty() : tensor<1x10xf32> loc(#loc11)
+    %7 = "ttir.matmul"(%5, %arg2, %6) <{operand_constraints = [#any_device, #any_device, #any_device]}> : (tensor<1x256xf32>, tensor<256x10xf32>, tensor<1x10xf32>) -> tensor<1x10xf32> loc(#loc11)
+    // CHECK: %{{.+}} = "ttnn.matmul"([[I1:%.+]], [[I2:%.+]], [[O4:%.+]]) {{.+}} -> tensor<1x10xf32, {{.+}}>
+    // CHECK: "ttnn.dealloc"([[I2]]) : (tensor<256x10xf32, {{.+}}>) -> ()
+    // CHECK: "ttnn.dealloc"([[O3]]) : (tensor<1x256xf32,{{.+}}>) -> ()
+    %8 = tensor.empty() : tensor<1x10xf32> loc(#loc12)
+    %9 = "ttir.add"(%7, %arg1, %8) <{operandSegmentSizes = array<i32: 2, 1>, operand_constraints = [#any_device, #any_device, #any_device]}> : (tensor<1x10xf32>, tensor<1x10xf32>, tensor<1x10xf32>) -> tensor<1x10xf32> loc(#loc12)
```

Comment:
nit: Add one more test case to test that we are deallocating properly when op result is used as dps operand of another op

Reply:
TTNN.empty already covers this case, I agree depth is 2(1+1 jump), only single loop of DPS init iteration skips but it's whats currently supported.